plugins/message-type/src/lib.rs

```diff
@@ -102,20 +102,40 @@ impl Plugin<Message> for MsgType {
         let fname = network.and_then(|net| net.file_name());
         // message that will be returned
         let mut resp = util::new_msg(req, server_id, sname, fname);
-
-        // if there is a server identifier it must match ours
-        if matches!(req.opts().get(OptionCode::ServerIdentifier), Some(DhcpOption::ServerIdentifier(id)) if *id != server_id && !id.is_unspecified())
+        let server_id_override = util::get_server_identifier_override(req.opts());
+
+        // determine if we should use the server identifier override as the server identifier in the response message
+        let use_server_id_override =
+            if let (Some(&override_id), Some(&DhcpOption::ServerIdentifier(server_identifier))) = (
+                server_id_override,
+                req.opts().get(OptionCode::ServerIdentifier),
+            ) {
+                override_id == server_identifier
+            } else {
+                false
+            };
+        // if there is a server identifier it must match either the server identifier override address or ours
+        if matches!(req.opts().get(OptionCode::ServerIdentifier), Some(DhcpOption::ServerIdentifier(id)) if *id != server_id && !id.is_unspecified() && !use_server_id_override)
         {
-            debug!(?server_id, "server identifier in msg doesn't match");
+            debug!(?server_id, "server identifier in msg doesn't match server address or server identifier override");
             return Ok(Action::NoResponse);
         }
         if req.opcode() == Opcode::BootReply {
             debug!("BootReply not supported");
             return Ok(Action::NoResponse);
         }
-        // add server id to response
+
+        // determine which server identifier to use in the response message
+        let resp_server_id = if use_server_id_override {
```

Comment:
I wonder if we can rewrite this in such a way that the `unwrap` is not necessary. While it currently won't panic, future refactors can change that.

I'm not 100% clear on why kea's impl doesn't actually replace the server id sent to the client when an override is present though. Do you know?

Reply:
This has been refactored into a single match statement that performs the checks and assigns the values to `resp_server_id`. This could also be written as a slightly longer but more explicit `if let`, i.e.

```rust
let (use_server_id_override, resp_server_id) = 
    if let (Some(override_id), Some(&DhcpOption::ServerIdentifier(msg_server_id))) = (
        server_id_override,
        req.opts().get(OptionCode::ServerIdentifier),
    ) {
        if override_id == msg_server_id {
            (true, override_id)
        } else {
            (false, cfg_server_id)
        }
    } else {
        (false, cfg_server_id)
    };
```

With respect to Kea's implementation of RFC 5107, this involves conjecture on my part, but from what I can tell it is a combination of ISC being optimistic about the pace at which people would switch from IPv4 to IPv6 which has [support for relay interactions](https://datatracker.ietf.org/doc/html/rfc3315#section-7) and them concluding that simply comparing the message server id to the override id to issue a response (without storing RAI) was good enough as: ["the functionality handles expected use cases"](https://kea.readthedocs.io/en/kea-2.2.0/arm/dhcp4-srv.html).

Reply:
Maybe this is a good place to leverage the type system and create an `enum` where `ServerId` is one of two variants?
```
enum ServerId<'a> {
   RelayOverride(&'a [u8]),
   Id(&'a [u8]) // maybe there is a better name for this variant
}
```
or something along those lines. To make condition checking easier one could implement an `is_override(&self) -> bool` method. 

I haven't fully thought that through though, so if it doesn't simplify the code we don't need to pursue it.

Reply:
On the face of it, I am quite fond of this idea. I will play around with this a bit. Thank you! 🙏

Reply:
So I took this idea and ran with it a bit. I moved the logic that determines which server id should be used in the response into its own enum (`RespServerId`) which also can be `None` in the event that we should not respond (the message is not addressed to us, nor is it unspecified).

This has the net effect of simplifying the logic in `MsgType::handle` while maintaining the former behavior. I would be happy to make further adjustments to this as well if it would be more effective or if I've missed anything.

Reply:
Yeah, that seems like an improvement. One thing I would suggest is maybe making `new` take `&req` so we can move some of the `let` declarations out of `handle`, and take `get` and moving it to the `if let Some(..) = resp_server_id.get()`

We can go ahead and remove the draft on this afterwards if you like. Thanks for humoring the comments and changes.

Reply:
This sounds sensible to me. We now take `&req` as an arg in `RespServerId::new` and move the `let` declarations for the `msg_server_id_opt` and `server_id_override` out of `handle`. I have also added RFC 5107 to the readme.

Thank you very much for your time and constructive feedback! 🙏